pkg/databind/internal/secrets/http_test.go

```diff
@@ -0,0 +1,351 @@
+
+package secrets
+
+import (
+	"bytes"
+	"crypto/rand"
+	"crypto/rsa"
+	"crypto/tls"
+	"crypto/x509"
+	"crypto/x509/pkix"
+	"encoding/json"
+	"encoding/pem"
+	"fmt"
+	"math/big"
+	"net"
+	. "net/http"
+	"net/http/httptest"
+	"os"
+	"path"
+	"regexp"
+	"testing"
+	"time"
+
+	"github.com/newrelic/infrastructure-agent/pkg/databind/pkg/data"
+)
+
+type certs struct {
+	CACert		string
+	CAKey		string
+	ServerCert	string
+	ServerKey	string
+	ClientCert	string
+	ClientKey	string
+}
+
+var certPaths	certs
+
+func TestClientCertHttpRequest(t *testing.T) {
+	ts, err := newLocalHttpsTestServer(
+		`{ "message": "Hello, world" }`, 200)
+	if err != nil {
+		t.Errorf("failed to start TLS servr: %s", err)
+	}
+
+	defer ts.Close()
+	HTTP := &http{
+		URL:     ts.URL,
+		TLSConfig: (tlsConfig{
+			Enable: true,
+			InsecureSkipVerify: false,
+			Ca: certPaths.CACert,
+			ClientCert: certPaths.ClientCert,
+			ClientKey: certPaths.ClientKey,
+		}),
+		Headers: make(map[string]string),
+	}
+
+	dt, err := httpRequest(HTTP, "GET", nil)
+	if err != nil {
+		t.Errorf("call failed: %s", err)
+	}
+
+	smap := data.InterfaceMap{}
+	if err := json.Unmarshal(dt, &smap); err != nil {
+		t.Errorf("unable to decode response: %s", err)
+	}
+
+	if smap["message"] != "Hello, world" {
+		t.Errorf("expected message, got %v", smap)
+	}
+}
+
+func TestClientCertHttpRequestFails(t *testing.T) {
+	ts, err := newLocalHttpsTestServer(
+		`{ "message": "Hello, world" }`, 200)
+	if err != nil {
+		t.Errorf("failed to start TLS servr: %s", err)
+	}
+
+	defer ts.Close()
+	HTTP := &http{
+		URL:     ts.URL,
+		TLSConfig: (tlsConfig{
+			Enable: true,
+			InsecureSkipVerify: false,
+			Ca: certPaths.CACert,
+		}),
+		Headers: make(map[string]string),
+	}
+
+	_, err = httpRequest(HTTP, "GET", nil)
+	if err == nil {
+		t.Errorf("call should have returned error but did not")
+	}
+
+	match, _ := regexp.MatchString("remote error: tls:", err.Error())
+	if !match {
+		t.Errorf("error did not match expected pattern: %s", err.Error())
+	}
+}
+
+func newLocalHttpsTestServer(response string, rc int) (*httptest.Server, error) {
+	ts := httptest.NewUnstartedServer(HandlerFunc(func(w ResponseWriter, r *Request) {
+		w.WriteHeader(rc)
+		w.Write([]byte(response))
+	}))
+
+    cert, err := tls.LoadX509KeyPair(certPaths.ServerCert, certPaths.ServerKey)
+    if err != nil {
+        return nil, err
+    }
+
+	certpool := x509.NewCertPool()
+	
+	pem, err := os.ReadFile(certPaths.CACert)
+	if err != nil {
+		return nil, fmt.Errorf("Failed to read certificate authority: %v", err)
+	}
+
+	if !certpool.AppendCertsFromPEM(pem) {
+		return nil, fmt.Errorf("Can't parse certificate authority")
+	}
+
+    ts.TLS = &tls.Config{
+		Certificates: []tls.Certificate{cert},
+		ClientAuth:   tls.RequireAndVerifyClientCert,
+		ClientCAs:    certpool,
+	}
+
+    ts.StartTLS()
+    return ts, nil
+}
+
+func TestMain(m *testing.M) {
+	err := setup()
+	if err != nil {
+		os.Exit(-1)
+	}
+
+    exitVal := m.Run()
+
+    teardown()
```

Comment:
What do you think of moving the `teardown()` function call at the top with a defer statement? That will assure to remove the files if there is a failure in the setup.


```suggestion
        defer teardown()
	err := setup()
	if err != nil {
		os.Exit(-1)
	}

    exitVal := m.Run()

    
```